plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataBuilder.java

```diff
@@ -0,0 +1,334 @@
+ *
+ * @since 3.0
+ */
+public class MetadataBuilder
+    extends ComponentSupport
+{
+  private final VersionScheme versionScheme;
+
+  private String groupId;
+
+  private String artifactId;
+
+  private String baseVersion;
+
+  // G level
+
+  private final List<Plugin> plugins;
+
+  // A level
+
+  private final TreeSet<Version> baseVersions;
+
+  // V level
+
+  private final Map<String, VersionCoordinates> latestVersionCoordinatesMap;
+
+  private VersionCoordinates latestVersionCoordinates;
+
+  public MetadataBuilder() {
+    this.versionScheme = new GenericVersionScheme();
+    // G
+    this.plugins = Lists.newArrayList();
+    // A
+    this.baseVersions = Sets.newTreeSet();
+    // V
+    this.latestVersionCoordinatesMap = Maps.newHashMap();
+  }
+
+  // -----------------------------------
+  // context
+
+  public String getGroupId() {
+    return groupId;
+  }
+
+  private boolean setGroupId(final String groupId) {
+    checkNotNull(groupId, "groupId");
+    if (Objects.equals(groupId, this.getGroupId())) {
+      return false;
+    }
+    this.groupId = groupId;
+    this.artifactId = null;
+    this.baseVersion = null;
+    return true;
+  }
+
+  public String getArtifactId() {
+    return artifactId;
+  }
+
+  private boolean setArtifactId(final String artifactId) {
+    checkState(groupId != null, "groupId == null");
+    checkNotNull(artifactId, "artifactId");
+    if (Objects.equals(artifactId, this.getArtifactId())) {
+      return false;
+    }
+    this.artifactId = artifactId;
+    this.baseVersion = null;
+    return true;
+  }
+
+  public String getBaseVersion() {
+    return baseVersion;
+  }
+
+  private boolean setBaseVersion(final String baseVersion) {
+    checkState(groupId != null, "groupId == null");
+    checkState(artifactId != null, "artifactId == null");
+    checkNotNull(baseVersion, "baseVersion");
+    if (Objects.equals(baseVersion, this.getBaseVersion())) {
+      return false;
+    }
+    this.baseVersion = baseVersion;
+    return true;
+  }
+
+  // -----------------------------------
+  // groupId
+
+  public boolean onEnterGroupId(final String groupId) {
+    if (setGroupId(groupId)) {
+      plugins.clear();
+      return true;
+    }
+    return false;
+  }
+
+  @Nullable
+  public Maven2Metadata onExitGroupId() {
+    checkState(getGroupId() != null, "groupId");
+    if (plugins.isEmpty()) {
+      log.debug("No plugins in group: {}:{}", getGroupId());
+      return null;
+    }
+    return Maven2Metadata.newGroupLevel(DateTime.now(), getGroupId(), plugins);
+  }
+
+  public boolean addPlugin(final String prefix, final String artifactId, final String name) {
+    final Plugin plugin = new Plugin(artifactId, prefix, name);
+    for (Plugin existing : plugins) {
+      if (Objects.equals(plugin.getArtifactId(), existing.getArtifactId())
+          && Objects.equals(plugin.getPrefix(), existing.getPrefix())) {
+        existing.setName(plugin.getName()); // update name, else is OK
```

Comment:
I basically reimplemented what NX2 does. The plugin list is keyed by (artifactId, prefix) pair only, name is auxiliary information only (and is not part of uq key). Frankly, it bothers me too, as this makes Plugin mutable, and looks awkward. Will redo this part.